statistics/histogram.go

```diff
@@ -1186,14 +1186,70 @@ func (idx *Index) GetRowCount(sc *stmtctx.StatementContext, indexRanges []*range
 		}
 		l := types.NewBytesDatum(lb)
 		r := types.NewBytesDatum(rb)
-		totalCount += idx.BetweenRowCount(l, r)
 		lowIsNull := bytes.Equal(lb, nullKeyBytes)
 		if (idx.outOfRange(l) && !(isSingleCol && lowIsNull)) || idx.outOfRange(r) {
 			totalCount += outOfRangeEQSelectivity(outOfRangeBetweenRate, modifyCount, int64(idx.TotalRowCount())) * idx.TotalRowCount()
 		}
 		if isSingleCol && lowIsNull {
 			totalCount += float64(idx.NullCount)
 		}
+		// Due to the limitation of calcFraction and convertDatumToScalar, the histogram actually won't estimate anything.
+		// If the first column's range is point.
+		if rangePosition := GetOrdinalOfRangeCond(sc, indexRange); rangePosition > 0 && idx.StatsVer == Version2 && coll != nil {
+			tmpRan := []*ranger.Range{
+				{
+					LowVal:  make([]types.Datum, 1),
+					HighVal: make([]types.Datum, 1),
+				},
+			}
+			colsIDs := coll.Idx2ColumnIDs[idx.ID]
+			singleColumnEstResults := make([]float64, 0, len(indexRange.LowVal))
+			// The following codes are the same with SQLServer's CSelCalcCombineFilters_ExponentialBackoff Calculator. It works like:
+			//   1. Calc the selectivity of each column.
+			//   2. Sort them and choose the first 4 most selective filter and the corresponding selectivity is sel_1, sel_2, sel_3, sel_4 where i < j => sel_i < sel_j.
+			//   3. The final selectivity would be sel_1 * sel_2^{1/2} * sel_3^{1/4} * sel_4^{1/8}.
+			// This calculation reduced the independce assumption and can work well better than it.
+			for i := 0; i < len(indexRange.LowVal) && i < len(colsIDs); i++ {
+				tmpRan[0].LowVal[0] = indexRange.LowVal[i]
+				tmpRan[0].HighVal[0] = indexRange.HighVal[i]
+				if i == len(indexRange.LowVal)-1 {
+					tmpRan[0].LowExclude = indexRange.LowExclude
+					tmpRan[0].HighExclude = indexRange.HighExclude
+				}
+				colID := colsIDs[i]
+				var count float64
+				if anotherIdxID, ok := coll.ColID2IdxID[colID]; ok && anotherIdxID != idx.ID {
+					count, err = coll.GetRowCountByIndexRanges(sc, anotherIdxID, tmpRan)
```

Comment:
Is there any risk to cause an infinite recursion?
For example when estimating `where idx_a > 1 and idx_a < 10`, the stack would be:
```
Index.GetRowCount
HistColl.GetRowCountByIndexRanges
Index.GetRowCount
HistColl.GetRowCountByIndexRanges
Index.GetRowCount
...
HistColl.GetRowCountByIndexRanges
Index.GetRowCount
HistColl.GetRowCountByIndexRanges
Index.GetRowCount
```

Reply:
Yes, so we check `anotherIdxID != idx.ID` here. This will not visit the same index again.